Compiler should suggest expect/unwrap before the suggestion to remove the last method call

### Code

```Rust
fn main() {
    let abs: i32 = 3i32.checked_abs();
}
```


### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:2:20
  |
2 |     let abs: i32 = 3i32.checked_abs();
  |              ---   ^^^^^^^^^^^^^^^^^^ expected `i32`, found `Option<i32>`
  |              |
  |              expected due to this
  |
  = note: expected type `i32`
             found enum `Option<i32>`
help: try removing the method call
  |
2 -     let abs: i32 = 3i32.checked_abs();
2 +     let abs: i32 = 3i32;
  |

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to previous error
```


### Desired output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
 --> src/main.rs:2:20
  |
2 |     let abs: i32 = 3i32.checked_abs();
  |              ---   ^^^^^^^^^^^^^^^^^^ expected `i32`, found `Option<i32>`
  |              |
  |              expected due to this
  |
  = note: expected type `i32`
             found enum `Option<i32>`
help: consider using `Option::expect` to unwrap the `Option<i32>` value, panicking if the value is an `Option::None`
   |
2  |     let abs: i32 = 3i32.checked_abs().expect("REASON");
   |                                      +++++++++++++++++

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to previous error
```


### Rationale and extra context

#105872 in Dec '22 introduced a cool new feature to suggest removing the last method call if it would alleviate a type mismatch error. This is particularly using for people wrestling with `my_str.to_string()` issues.

#116841 in Oct '23 introduced another useful feature to suggest to users that they might consider `expect`ing an `Option` or `Result` type in order to solve a type mismatch. Also very cool, and conveniently by the same author @chenyukang.

Unfortunately, in cases where a method `T -> Option<T>` is called, the previous check takes precedence, and the compiler suggests removing that trailing method to get the code to compile. This isn't generally that useful as the expect suggestion, and I can't think of an example where it's a better one. If you can think of one though, please put it below for me to test!

#116738 surfaced this with `Path.parent()` but it's more easily demonstrated with ints.

I can see two ways to fix this:
1. Run the trailing method check first, but ignore `Result` and `Option` types
2. Reorder the checks so the option type is checked first. Note that the method check has to be high in the chain as the motive is to quickly suggest removing a call, before adding a whole load of new ones. This means the `expect` check has to be one of the first

I am going to work on solving it via method 2, as it feels cleaner that the method check doesn't have to know about the `Option` and `Result` check.

If anyone has strong feelings on this, please let me know!

Playground [here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=55229f3d8e08048d02a12374c9ca03e6)

### Other cases

_No response_

### Anything else?

_No response_

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"3tilley"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/demand.rs
+++ b/compiler/rustc_hir_typeck/src/demand.rs
@@ -43,6 +43,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         // Use `||` to give these suggestions a precedence
         let suggested = self.suggest_missing_parentheses(err, expr)
+            || self.suggest_missing_unwrap_expect(err, expr, expected, expr_ty)
             || self.suggest_remove_last_method_call(err, expr, expected)
             || self.suggest_associated_const(err, expr, expected)
             || self.suggest_deref_ref_or_into(err, expr, expected, expr_ty, expected_ty_expr)
@@ -58,8 +59,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             || self.suggest_into(err, expr, expr_ty, expected)
             || self.suggest_floating_point_literal(err, expr, expected)
             || self.suggest_null_ptr_for_literal_zero_given_to_ptr_arg(err, expr, expected)
-            || self.suggest_coercing_result_via_try_operator(err, expr, expected, expr_ty)
-            || self.suggest_missing_unwrap_expect(err, expr, expected, expr_ty);
+            || self.suggest_coercing_result_via_try_operator(err, expr, expected, expr_ty);
 
         if !suggested {
             self.note_source_of_type_mismatch_constraint(

--- a/tests/ui/suggestions/issue-117669.rs
+++ b/tests/ui/suggestions/issue-117669.rs
@@ -1,0 +1,4 @@
+fn main() {
+    let abs: i32 = 3i32.checked_abs();
+    //~^ ERROR mismatched types
+}

--- a/tests/ui/suggestions/issue-117669.stderr
+++ b/tests/ui/suggestions/issue-117669.stderr
@@ -1,0 +1,18 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-117669.rs:2:20
+   |
+LL |     let abs: i32 = 3i32.checked_abs();
+   |              ---   ^^^^^^^^^^^^^^^^^^ expected `i32`, found `Option<i32>`
+   |              |
+   |              expected due to this
+   |
+   = note: expected type `i32`
+              found enum `Option<i32>`
+help: consider using `Option::expect` to unwrap the `Option<i32>` value, panicking if the value is an `Option::None`
+   |
+LL |     let abs: i32 = 3i32.checked_abs().expect("REASON");
+   |                                      +++++++++++++++++
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

